pkg/compact/compact.go

```diff
@@ -1020,6 +1029,39 @@ func (cg *Group) areBlocksOverlapping(include *metadata.Meta, exclude ...*metada
 	return nil
 }
 
+func (cg *Group) removeOverlappingBlocks(ctx context.Context, toCompact []*metadata.Meta, dir string, blockDeletableChecker BlockDeletableChecker) error {
+	if len(toCompact) == 0 {
+		return nil
+	}
+	kept := toCompact[0]
+	for _, m := range toCompact {
+		if m.MinTime < kept.MinTime && m.MaxTime > kept.MaxTime {
+			level.Warn(cg.logger).Log("msg", "found overlapping block in plan that are not the first",
+				"first", kept.String(), "block", m.String())
+			kept = m
+		} else if m.MinTime < kept.MinTime || m.MaxTime > kept.MaxTime {
+			return halt(errors.Errorf("found partially overlapping block: %s vs %s", m.String(), kept.String()))
+		}
+	}
+	for _, m := range toCompact {
+		if m.ULID.Compare(kept.ULID) == 0 {
+			level.Info(cg.logger).Log("msg", "skip the longest overlapping block", "block", m.String(),
+				"level", m.Compaction.Level, "source", m.Thanos.Source, "labels", m.Thanos.Labels)
+			continue
+		}
+		cg.blocksOverlapped.Inc()
+		if err := os.RemoveAll(filepath.Join(dir, m.ULID.String())); err != nil {
+			return errors.Wrapf(err, "remove old block dir %s", m.String())
+		}
+		if blockDeletableChecker.CanDelete(cg, m.ULID) {
+			level.Warn(cg.logger).Log("msg", "deleting overlapping block", "block", m.String(),
+				"level", m.Compaction.Level, "source", m.Thanos.Source, "labels", m.Thanos.Labels)
+			return block.Delete(ctx, cg.logger, cg.bkt, m.ULID)
+		}
```

Comment:
This is not future-proof. If upstream adds another condition (not included in `blockDeletableChecker`) for a block deletion, this deletion logic would be broken. We need to careful when pull in upstream changes.

Reply:
makes sense, I am thinking to remove the block checker to force deletion here.